staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go

```diff
@@ -1113,6 +1113,82 @@ type APIResource struct {
 	StorageVersionHash string `json:"storageVersionHash,omitempty" protobuf:"bytes,10,opt,name=storageVersionHash"`
 }
 
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Discovery API Types
+// APIGroupList is a list of APIGroup, to allow clients to discover the API at
+// /apis.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type DiscoveryAPIGroupList struct {
+	TypeMeta `json:",inline"`
+	// groups is a list of APIGroup.
+	// +listType=map
+	// +listMapKey=name
+	Groups []DiscoveryAPIGroup `json:"groups" protobuf:"bytes,1,rep,name=groups"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// APIGroup contains the name, the supported versions, and the preferred version
+// of a group.
+type DiscoveryAPIGroup struct {
+	TypeMeta `json:",inline"`
+	// name is the name of the group.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// versions are the versions supported in this group.
+	// This will be sorted in descending order based on the preferred version
```

Comment:
@lavalamp do you know of any case in the wild, where production clusters get versions of things installed that are not in the v1, v2beta1, v1alpha3, style?

Reply:
I vaguely recall one person wanted to do that, but I don't recall any cases of actually doing that.

Reply:
The Azure Service Operator uses a somewhat novel variant of the typical v1betaN pattern. Pretty sure it will still be compatible here though: https://azure.github.io/azure-service-operator/reference/